src/java.base/share/classes/jdk/internal/foreign/SegmentBulkOperations.java

```diff
@@ -0,0 +1,271 @@
+            for (; offset < limit; offset += 8) {
+                final long v = SCOPED_MEMORY_ACCESS.getLong(src.sessionImpl(), src.unsafeGetBase(), src.unsafeGetOffset() + srcOffset + offset);
+                SCOPED_MEMORY_ACCESS.putLong(dst.sessionImpl(), dst.unsafeGetBase(), dst.unsafeGetOffset() + dstOffset + offset, v);
+            }
+            int remaining = (int) size - offset;
+            // 0...0X00
+            if (remaining >= 4) {
+                final int v = SCOPED_MEMORY_ACCESS.getInt(src.sessionImpl(), src.unsafeGetBase(),src.unsafeGetOffset() + srcOffset + offset);
+                SCOPED_MEMORY_ACCESS.putInt(dst.sessionImpl(), dst.unsafeGetBase(), dst.unsafeGetOffset() + dstOffset + offset, v);
+                offset += 4;
+                remaining -= 4;
+            }
+            // 0...00X0
+            if (remaining >= 2) {
+                final short v = SCOPED_MEMORY_ACCESS.getShort(src.sessionImpl(), src.unsafeGetBase(), src.unsafeGetOffset() + srcOffset + offset);
+                SCOPED_MEMORY_ACCESS.putShort(dst.sessionImpl(), dst.unsafeGetBase(), dst.unsafeGetOffset() + dstOffset + offset, v);
+                offset += 2;
+                remaining -=2;
+            }
+            // 0...000X
+            if (remaining == 1) {
+                final byte v = SCOPED_MEMORY_ACCESS.getByte(src.sessionImpl(), src.unsafeGetBase(), src.unsafeGetOffset() + srcOffset + offset);
+                SCOPED_MEMORY_ACCESS.putByte(dst.sessionImpl(), dst.unsafeGetBase(), dst.unsafeGetOffset() + dstOffset + offset, v);
+            }
+            // We have now fully handled 0...0X...XXXX
+        } else {
+            // For larger sizes, the transition to native code pays off
+            SCOPED_MEMORY_ACCESS.copyMemory(src.sessionImpl(), dst.sessionImpl(),
+                    src.unsafeGetBase(), src.unsafeGetOffset() + srcOffset,
+                    dst.unsafeGetBase(), dst.unsafeGetOffset() + dstOffset, size);
+        }
+    }
+
+    @ForceInline
+    public static long mismatch(AbstractMemorySegmentImpl src, long srcFromOffset, long srcToOffset,
+                                AbstractMemorySegmentImpl dst, long dstFromOffset, long dstToOffset) {
+        final long srcBytes = srcToOffset - srcFromOffset;
+        final long dstBytes = dstToOffset - dstFromOffset;
+        src.checkAccess(srcFromOffset, srcBytes, true);
+        dst.checkAccess(dstFromOffset, dstBytes, true);
+
+        final long bytes = Math.min(srcBytes, dstBytes);
+        final boolean srcAndDstBytesDiffer = srcBytes != dstBytes;
+
+        if (bytes == 0) {
+            return srcAndDstBytesDiffer ? 0 : -1;
+        } else if (bytes < NATIVE_THRESHOLD_MISMATCH) {
+            return mismatch(src, srcFromOffset, dst, dstFromOffset, 0, (int) bytes, srcAndDstBytesDiffer);
+        } else {
+            long i;
+            if (SCOPED_MEMORY_ACCESS.getByte(src.sessionImpl(), src.unsafeGetBase(), src.unsafeGetOffset() + srcFromOffset) !=
```

Comment:
I looked at this the other day, and I couldn't immediately tell whether this test is needed or not - shouldn't it be covered by the subsequent loop? Is it a shortcut (but only for the first element) - how much does that buy really?

Reply:
I think this is an escape hatch that may trigger for random contents. Depending on the use case, this may or may not pay off. Maybe we should define an input distribution for which performance should be tuned against.